landscape/lib/lsb_release.py

```diff
@@ -1,26 +1,58 @@
-"""Get information from /etc/lsb_release."""
+"""Get information from /usr/bin/lsb_release."""
+import os
+from subprocess import DEVNULL, CalledProcessError, check_output
 
-LSB_RELEASE_FILENAME = "/etc/lsb-release"
-LSB_RELEASE_INFO_KEYS = {"DISTRIB_ID": "distributor-id",
-                         "DISTRIB_DESCRIPTION": "description",
-                         "DISTRIB_RELEASE": "release",
-                         "DISTRIB_CODENAME": "code-name"}
+LSB_RELEASE = "/usr/bin/lsb_release"
+LSB_RELEASE_FILENAME = "/etc/lsb_release"
+LSB_RELEASE_FILE_KEYS = {
+    "DISTRIB_ID": "distributor-id",
+    "DISTRIB_DESCRIPTION": "description",
+    "DISTRIB_RELEASE": "release",
+    "DISTRIB_CODENAME": "code-name",
+}
 
 
-def parse_lsb_release(lsb_release_filename):
-    """Return a C{dict} holding information about the system LSB release.
+def parse_lsb_release(lsb_release_filename=None):
+    """
+    Returns a C{dict} holding information about the system LSB release.
+    Reads from C{lsb_release_filename} if it exists, else calls
+    C{LSB_RELEASE}
+    """
+    if lsb_release_filename and os.path.exists(lsb_release_filename):
+        return parse_lsb_release_file(lsb_release_filename)
+
+    try:
+        lsb_info = check_output([LSB_RELEASE, "-as"], stderr=DEVNULL)
+    except CalledProcessError:
```

Comment:
I tried this and I am getting FileNotFound error instead of CalledProcessError if lsb-release is not installed on the system so it doesn't fall back to reading the file

Reply:
Right. overlooked that, will fix.

Reply:
I think I would prefer to using one method, probably the command line. It keeps it simple to use one method and also in debian falling back to the file is not going to work, (unless you use the /etc/os-release but then it's gonna be different keys for it). lsb-release should be installed in ubuntu and I'm fine with the release information missing from server if this is not installed, but I can also understand having it be foolproof in ubuntu  so your call

Reply:
I'm going to respectfully decline restricting it to one method, but not only for "foolproofness". Currently a lot of the application code passes in `lsb_release_filename`, so not having it handle the file-reading case necessitates a _lot_ of additional touches to the code-base.